WIP: Feature - language registration/discovery

igordejanovic opened the pull request:
This PR is WIP on #175 

@goto40 @danixeee I'm publishing PR to get early feedback as I go along.

## TODO Checklist

- [x] registration API tests (@igordejanovic)
- [x] metamodel `reference` tests (@igordejanovic)
- [x] remove `visualize` command in favor of `generate` (@igordejanovic )
- [x] registration API docs (@igordejanovic)
- [x] meta-model `reference` docs (@igordejanovic)
- [x] `MetaModelProvider` removal and scoping tests update (@goto40, #178)
- [x] Scoping/multi-meta-model docs update (@goto40)
- [x] Allow passing in custom params to generators (@igordejanovic , see [this](https://github.com/textX/textX/issues/175#issuecomment-481322984))

## Code review checklist

- [ ] Pull request represents a single change (i.e. not fixing disparate/unrelated things in a single PR)
- [ ] Title summarizes what is changing
- [ ] Commit messages are meaningful (see [this][commit messages] for details)
- [ ] Tests have been included and/or updated
- [ ] Docstrings have been included and/or updated, as appropriate
- [ ] Standalone docs have been updated accordingly
- [ ] Changelog(s) has/have been updated, as needed (see `CHANGELOG.md`).


[commit messages]: https://chris.beams.io/posts/git-commit/

igordejanovic committed:
Fix model docs - `get_metamodel` description

igordejanovic committed:
Add named tuples for language/generator registration

igordejanovic committed:
Register textX language

igordejanovic committed:
Add `list_generators` and `list_languages` commands. Reorganize CLI

igordejanovic committed:
Simplify list generators/languages command

igordejanovic committed:
Improve registration API

igordejanovic committed:
Check if multiple registration of language/generator

igordejanovic committed:
Add API for programmatic registration of languages/generators

igordejanovic committed:
Expose registration API

igordejanovic committed:
Improve and extend language registration API

igordejanovic committed:
Treat textX language as any other language

igordejanovic committed:
Rework `check` command

igordejanovic committed:
Move public API function to the top

igordejanovic committed:
Use file patterns instead of file extensions

igordejanovic committed:
Add ability for `check` command to accept multiple model files

igordejanovic committed:
Add grammar support for language qualified references

igordejanovic committed:
Change grammar to use `reference` keyword for referencing languages

igordejanovic committed:
Improve error reporting for textX semantic errors

igordejanovic committed:
Revert `language_name` separate handling in references

igordejanovic committed:
Change `class_name` match to accept qualified identifiers

igordejanovic committed:
Collect referenced languages/meta-models

igordejanovic committed:
Change namedtuples for registration to classes

igordejanovic committed:
Add functions for clearing of registrations

igordejanovic committed:
Change match for language_name to allow for `.` and `-`

igordejanovic committed:
Change meta-class reference resolving to use new API

igordejanovic committed:
Update test_multi_metamodel_refs tests to use the new API

igordejanovic committed:
Remove `render_all_grammars` example from tests

igordejanovic committed:
Rework registration API. Generators + simplification.

igordejanovic committed:
Implement `generate` command

igordejanovic committed:
Improve `check` command

igordejanovic committed:
PEP-8 fixes

igordejanovic committed:
Add script that install test projects and run all tests

danixeee commented:
@igordejanovic Should you add package `textx.cli` to [setup.py](https://github.com/textX/textX/blob/feature/language-registration/setup.py#L49) ? CLI is not working when I install _textX_ with `pip install .` .

igordejanovic committed:
Test projects based on https://github.com/textX/textx-multi-metamodel-example

igordejanovic committed:
Add initial tests for discovery

igordejanovic committed:
Add tests for `check` command

igordejanovic committed:
PEP-8 and variable name fix for `check` tests

igordejanovic committed:
Tests for `generate` command

igordejanovic commented:
@danixeee Yup. Good catch. Thanks.

igordejanovic committed:
Add `textx.cli` to setup.py packages

igordejanovic committed:
Split installation and testing script

igordejanovic commented:
Ok, this branch is getting into shape.

@goto40 I would really appreciate some help here to migrate scoping to the new registration API. The  idea is to replace `MetaModelProvider`. Please see changes in `tests/functional/test_metamodel/` tests and project examples in `tests/functional/registration/projects`. Those project are based on [multimetamodel example](https://github.com/textX/textx-multi-metamodel-example) so the example should be updated accordingly. You can branch your work of this feature branch and make PR against this  branch.

Check how `check` command is now agnostic whether you are checking model or meta-model (each .tx file is a model of textX language/meta-meta-model :) ).  But still you can use command like before, with unregistered language/meta-model from `.tx` file.

All textX commands are registered and treated in the same way. No magical commands. See `setup.py` for what is registered. I still have to migrate `visualize` to be just a plain generator for `dot` and `PlantUML` targets. This way we could move all those textX commands to a separate Python package/project if we decide to. Also, new commands can be added by third-party modules that will be no different from textX provided.

Still have some work to be done but I think this approach will be much cleaner and more consistent.

igordejanovic committed:
Hard fail on any issue during run of installation/testing script

igordejanovic commented:
I made scripts : 
- `install-all-dev.sh` - install textX in development mode and install all test projects
- `runtests.sh` - runs all tests, coverage and flake8

These scripts are now called from `.travis.yml`

igordejanovic committed:
Add installation of `flow_codegen` test project

igordejanovic commented:
@goto40 One important note: `referenced_metamodels` is removed from APIs (`metamodel_from_str`, `metamodel_from_file`) as it is now handled explicitly in grammars by `reference` keyword (e.g `reference types-dsl as t`).

goto40 commented:
I will have a look tomorrow... Wow. I am looking forward!

danixeee commented:
@igordejanovic I am writing a generator for VS Code extensions and I noticed that a language id is case-sensitive when I called `generate .../workflow.tx --target vscode` command. I guess that the generator should not be displayed when listing generators or _language_ field in `GeneratorDesc` can be case insensitive.

![image](https://user-images.githubusercontent.com/16227576/55366671-f9688300-54e9-11e9-9c3f-9d289e98d065.png)

![image](https://user-images.githubusercontent.com/16227576/55366751-67ad4580-54ea-11e9-9901-bfea228ba040.png)

P.S. Great work, it's really easy to use and understand! :)

goto40 commented:
Before i dive into details: the responsibility of the meta model provider was to deduce the meta model when importing a model file, based on its file extension.

Do you think this responsibility is still valid? If not, we must find another way to find the meta model for an importuri field...

igordejanovic commented:
@danixeee Yes, I don't think there is any point in making language and target names case-sensitive. Will change to insensitive matching.

@goto40 registration API now takes over all responsibilities of MetaModelProvider and adds more like generator registration/discovery, declarative registration of languages and generators in setup.py etc. For getting a metamodel from file or pattern use `metamodel_for_file`. See `textx.registration`. The API is exported in `textx` module and should be imported from there.

igordejanovic committed:
Change registration to case insensitive

igordejanovic committed:
Remove old unused lang. registration API

igordejanovic committed:
Rework and fix docstrings

goto40 committed:
first patch: allowed to use the own metamodel, when loading a file
with unknown extension

goto40 committed:
fixed missing "exists"; added a new function to get the metamodel by extension of a default metamodel (if no extension matches).

goto40 committed:
one more test fixed.

goto40 committed:
more tests fixed. flake8.

goto40 committed:
removed metamodelprovider

goto40 commented:
I removed the MetaModelProvider and fixed tests. See "major changes" in #178... we should discuss the changes.

**I really appreciate the new way of registering languages! This is a great addition to textx.**
However, I have no good feeling when clearing all languages (we had the same problem with the - now obsolete - MetaModelProvider)... Should we then at least automatically add some default langues again ("*.tx")?

I think we should add a default metamodel to the importURI scope provider somehow - or **allow to specify the default (or fixed)  metamodel to be used for an importURI statement**... At the moment I have no good idea for this - (the "import rule" must provide that information)... **Or maybe** we need a **metamodel specific metamodelprovider**?

igordejanovic committed:
Enable short form of registration without direct usage of `*Desc` classes

igordejanovic commented:
> However, I have no good feeling when clearing all languages (we had the same problem with the - now obsolete - MetaModelProvider)... Should we then at least automatically add some default langues again ("*.tx")?

That is perfectly fine. Clearing registration would bring the state of registry to the initial and the first access to the API will re-register declaratively registered languages (using the extension point). Since the textX language is registered as any other language it'll be always there. Check this for example:
```python
clear_language_registrations()
l = language_description('textx')
assert l.name = 'textX'
```

> I think we should add a default metamodel to the importURI scope provider somehow - or allow to specify the default (or fixed) metamodel to be used for an importURI statement... At the moment I have no good idea for this - (the "import rule" must provide that information)... Or maybe we need a metamodel specific metamodelprovider?

I'm repeating here a comment from #178 to be in context. I'm thinking about providing an optional syntax for `importURI` that could specify the language. For example:
```
import 'flow-dsl:/some/path/myflow.flow'
``` 
If the `<language>:` prefix exists than `metamodel_for_language` is called else `metamodel_for_file` is called to get the meta-model.

Another addition (to be consistent with the current `reference` keyword in textX language) is to enable:
```
import 'flow-dsl:/some/path/myflow.flow' as mf
```
And than we could use qualified names to `mf.SomeObject`. I think we have discussed this somewhere. Perhaps we could make a new TEP on this to collect information scattered around in one place?

igordejanovic committed:
Expose registration API for querying the descriptions

igordejanovic commented:
@goto40 I've added a TODO checklist at the top of the stuff need to be done before this is ready for merge.

igordejanovic commented:
> Or maybe we need a metamodel specific metamodelprovider?

That is something to think about. We might use now explicit meta-model references through `reference` keyword to deduce needed meta-model even if it is not explicitly specified in importURI. That is probably something for another PR, maybe post 2.x.

goto40 committed:
review comments included

goto40 committed:
Merge branch 'feature/language-registration-MetaModelProvider-replacement' of https://github.com/textX/textX into feature/language-registration-MetaModelProvider-replacement

goto40 committed:
used new style of register_languages

igordejanovic committed:
Rework command tests to use click testing API

igordejanovic committed:
Fix meta-model double instantiation during registration

igordejanovic committed:
Improve/fix registration example projects tests

igordejanovic committed:
PEP-8 fixes in tests

igordejanovic committed:
Omit six.py from test coverage

igordejanovic committed:
Fix inheritance for Python 2 in registration

goto40 commented:
Good work!

igordejanovic committed:
Add registration API tests

igordejanovic committed:
Improve tests for `check` command

igordejanovic committed:
Fix `generate` command

igordejanovic committed:
Improve `generate` command test coverage

igordejanovic committed:
Optimize check command to instantiate meta-model only once if possible

igordejanovic committed:
Improve `generate` command

igordejanovic committed:
Use special language `any` for generic generators

igordejanovic committed:
Allow for `any` language is language is not explicitly given

igordejanovic committed:
Replace `visualize` with generators

igordejanovic committed:
Add `generator` decorator for generator registration

igordejanovic committed:
Add test for `generator` decorator

igordejanovic commented:
@goto40  Thanks :)

igordejanovic committed:
PEP-8 fixes

igordejanovic committed:
Fix typos

igordejanovic committed:
Fix if grammar is given for generators language must be `any`

igordejanovic committed:
Add tests for generators/visualizators

igordejanovic committed:
Fix generators/vizualizators tests for Python 2

goto40 commented:
I would like to discuss, if we could not use the python package system instead of our custom language registration module?

Python is good in finding packages. I would like to reuse this functionality for finding languages.

What do both of you think about this? It should not be a big deal to change this at the current state?

goto40 commented:
For the user it just means to use the full package name or a unique substring of it (at the command line) to identify a language.

For the meta model designer it means not to have to think about a good language name, but about a good package name only.

igordejanovic commented:
Not sure that I understand but if you propose a dynamic load using a FQN of a Python package then I'm not convinced that it would be a good way to go. Here are a couple of reasons:
- If you are concerned about name clashing think that our languages would share the same namespace with all the Python packages. I generally don't think that we need some complex FQN naming scheme. If textX become so popular that we have so many languages around and the name clashing becomes a real problem... well that's the problem I would like to have :smile: 
- Python dynamic import has considerably changed from version to version so we would have to support all variants used in Python versions we support
- You would still need to register your module either declaratively or imperatively in order to be able to list it (`textx list-languages`) so I'm not really able to see any benefit in comparison to the current approach

goto40 commented:
OK. It was just an idea.

igordejanovic commented:
Sure, no problem. I like to discuss new ideas :)

goto40 commented:
Maybe we could allow to pass an optional list of LanguageDesc when creating a meta model.

This list could override global registration in case of conflicts... We could (or I could the next days) encapsulate the registration.py logic in an object. This object would then exist one time globally and could be created locally for individual meta models in an multi meta model context.

The global registration api is untouched. 

The local lookup/registration object could be accessible via _tx_attribute for, e.g., editors.

Local registration could be automatically forwarded to global registration, if no conflicts exist.

igordejanovic commented:
What problem would that solve?

igordejanovic commented:
If that would solve only the potential conflicts then I suggest to push that back to after 2.0 as it wont be a backward incompatible change so we could do that for 2.1. I would like to focus now on getting 2.0 out. I was about to do the release but figured out that this registration/discovery is important to be done ASAP to unblock @danixeee work on textX-LS.

If @danixeee can continue his work with this we can focus on updating tests and docs and releasing 2.0. After that we'll have more time to make a proper decisions and improvements. I also have some ideas on how to extend the possibilities of newly introduced `reference` keyword.

goto40 commented:
The use case I had in mind is the one you mentioned : a language l1 uses a language l2. If I install 'another' language l2 I break my language l1.

Local registration solves this problem. I can make a pull request after 2.0. Or we can discuss...

igordejanovic commented:
Ok. I suggest to move the discussion to a new issue and discuss the solution before the implementation.

danixeee commented:
> I was about to do the release but figured out that this registration/discovery is important to be done ASAP to unblock @danixeee work on textX-LS.

Yeah, I stopped with `textX-LS` for now, because this implementation allows us to create separate projects (generators) which will be used by language server, so I am just playing with generators for now. Once we address passing [custom arguments](https://github.com/textX/textX/issues/175#issuecomment-481322984) to generator, I will continue my work.

igordejanovic committed:
Ignore generated test file

igordejanovic committed:
Fix generate command test to not fail in case test output is missing

igordejanovic committed:
Rework textX CLI extensibility to be simpler and more flexible

Now you should register a Python function that accepts the top-level `textx`
command group and use that to register additional commands using `click`
decorators. In this way we use the same mechanism for both a plain command and
command group registration. In addition it is now possible to provide additional
parameters to `click` command parser which will be useful in advanced textX
commands.

igordejanovic committed:
Enable passing of arbitrary custom arguments to generators

igordejanovic commented:
@danixeee Generators now accept custom arguments. See the last commit.

danixeee commented:
@igordejanovic I have some problems getting custom arguments from CLI, can you take a look when you have chance?

Project: https://github.com/danixeee/textx-gen-coloring
Install: `pip install .`
Command: In Readme

Custom arguments dictionary is always empty [here](https://github.com/danixeee/textx-gen-coloring/blob/master/textx_gen_coloring/__init__.py#L30).

igordejanovic committed:
Fix passing custom arguments to generators

igordejanovic committed:
Merge branch 'master' into feature/language-registration

igordejanovic commented:
@danixeee It was a bug in custom arguments collecting. Fixed.

igordejanovic committed:
Add command for generating coverage report in html format

igordejanovic committed:
Add `language` decorator for register a new language

This is a convenience method to register a new language by decorating
function returning the language meta-model. Docstring is used as a
language description.

igordejanovic committed:
Update test projects to use new `language` decorator style

igordejanovic committed:
Update docs for textx command and visualization

igordejanovic committed:
Fix in language decorator to strip docstring

igordejanovic committed:
WIP: registration/discovery docs

igordejanovic committed:
Fix inconsistencies in registration API

igordejanovic committed:
Improve visualization docs

igordejanovic committed:
Finish registration/discovery docs

igordejanovic committed:
Fix docstring for `generate` command

danixeee commented:
I have implemented a basic [textmate generator](https://github.com/danixeee/textx-gen-coloring) for syntax highlighting. It works with the latest changes from this branch.

igordejanovic commented:
@danixeee Cool! Any docs on coloring DSL? How can we try the generator? I guess the generated highlighting file should be copied somewhere in VS Code installation or workspace or...?

I suggest to put a note in the README that this branch is needed if someone wants to test it before this PR is merged.

danixeee commented:
@igordejanovic I have implemented just default syntax coloring generator, which uses targeting language grammar and produces json file. It collects just strings (keywords) from a grammar.
That generator is called from [textx-gen-vscode](https://github.com/danixeee/textx-gen-vscode). For now, output is `.vsix` file (extension which can be installed). I will add `--install` flag, if we want to install the extension and delete `.vsix` file afterwards.

VS Code generator will be called from `textX-LS` in a subprocess, and after successful installation, if VS Code is open, `Reload` button will be displayed and user will have to reload the editor (there is no workaround for it AFAIK).

I will keep those generators in my repository until this PR is merged to master, then I will add test/docs and move them under `textX` organization.

Do you plan to include work on this PR in the next release?

igordejanovic commented:
@danixeee Sounds great.

> Do you plan to include work on this PR in the next release?

Yes, I'm planning to release 2.0.0 as soon as this PR is fully finished and merged.

igordejanovic commented:
@goto40 Regarding my previous comment, I see that you have ongoing work on `scope_provider_interactive_demo2`. Is it something you plan to merge before v2?

goto40 commented:
I want to add more unittests and refine things. I need some more days. Possibly we will need also to discuss things with @danixeee. 

However, the critical part was already merged with #186. My ongoing work should not break anything. So we can go ahead with textX 2.0.0...

igordejanovic committed:
language name should not allow dots

igordejanovic committed:
Rework multi-metamodel docs intro and the first use-case

igordejanovic committed:
Add tests for grammar `reference` statement

igordejanovic commented:
@goto40 I think I have finished this PR mostly. What remains is the update of scoping/multi-metamodel docs. I did a rework of introduction and first use-case in `multimetamodel.md`. 

Could you take a look when find some time and see if what I did is ok and could you update the rest of the docs?

goto40 commented:
I am happy to do it.

danixeee commented on textx/registration.py:
```diff
@@ -0,0 +1,312 @@
+    if languages is None:
+        language_descriptions()
+
+    if type(language_desc_or_name) is not LanguageDesc:
```
@igordejanovic Can you change this to check if `language_desc_or_name` is instance of `LanguageDesc`?
The reason is that we will need additional fields/methods for language description for `textX-LS`, and I thought it would be a good idea to avoid replacing similar logic in `textX-LS-core` and to allow `textX` to handle subclasses of `LanguageDesc`.

igordejanovic replied:
Sure. It sounds like a good idea.

igordejanovic replied:
Done, for both languages and generators.

danixeee commented on textx/registration.py:
```diff
@@ -0,0 +1,312 @@
+generators = None
+
+
+def language_descriptions():
```
A flag parameter like `reload_anyway` will be useful, as well for `generator_descriptions`.
I had to call `pkg_resources.WorkingSet().iter_entry_points(group='textx_languages')` in order to discover newly installed languages in already running process.

danixeee replied:
I see now that you have already added functions to clear languages and generators, sorry.

igordejanovic replied:
Yes. When the registration is cleared declaratively registered registrations gets reloaded with the next call to the API.

danixeee replied:
I think we still need to change to:

_pkg_resources`.WorkingSet()`.iter_entry_points(group='textx_languages')_

to reinitialize _entries_ field of working set, because `pip install ...` adds a new path on `sys.path`. Also, in `generator_descriptions` function.

https://github.com/pypa/setuptools/blob/8d4e4bc93e319275458dcc7381ec83d447eca770/pkg_resources/__init__.py#L564

igordejanovic replied:
You are right. Fixed. Thanks.

igordejanovic committed:
Instantiate a fresh `WorkingSet` on languages/generators discovery

- Necessary to account for `sys.path` changes after `pkg_resources` import.
- See this comment: https://github.com/textX/textX/pull/177#pullrequestreview-237566683

igordejanovic committed:
Enable usage of `LanguageDesc/GeneratorDesc` sub-classes in the registration API

goto40 committed:
Feature/language registration review2 (#187)

Adapted multi meta modeling to new registration api.

igordejanovic commented:
I think we are now in a good shape with this PR.
@goto40 @danixeee Is there anything on your end?

If everything looks ok I'll be working on merging this branch and releasing v2 tomorrow.

goto40 commented:
No more comments from my side. Looking forward to see v2!

danixeee commented:
Looks good to me as well. Check this out: https://github.com/textX/textX-LS/issues/13#issuecomment-493467118. :)

igordejanovic committed:
Merge branch 'master' into feature/language-registration

danixeee commented:
@igordejanovic Can we add project name to `LanguageDesc` and `GeneratorDesc` objects during entry point iteration?

igordejanovic commented:
@danixeee Yes, I think that would indeed be a valuable information. Done on `master` in commit  9d4fa74b7f77479f39774070c7f526d968078d41
